lib/engine/game/g_1812/game.rb

```diff
@@ -420,68 +420,102 @@ def trainless_penalty(entity)
 
+        PORT_COMPANIES_AND_BONUS_HEXES = {
+          p3_company: F3_PORT,
+          p8_company: F3_PORT,
+          p9_company: H9_PORT,
+          p12_company: G6_PORT,
+        }.freeze
+
         def g_train?(train)
           self.class::G_TRAINS.include?(train.name)
         end
 
-        def port_hexes
+        def port_hexes?(route)
           @port_hexes ||= PORT_HEXES.map { |coord| hex_by_id(coord) }
+          route.stops.map(&:hex).intersect?(@port_hexes)
+        end
+
+        def mine_hexes?(route)
+          @mine_hexes ||= MINE_HEXES.map { |coord| hex_by_id(coord) }
+          route.stops.map(&:hex).intersect?(@mine_hexes)
+        end
+
+        def north_hexes?(route)
+          @north_hexes ||= NORTH_HEXES.map { |coord| hex_by_id(coord) }
+          route.stops.map(&:hex).intersect?(@north_hexes)
+        end
+
+        def south_hexes?(route)
+          @south_hexes ||= MINE_HEXES.map { |coord| hex_by_id(coord) }
+          route.stops.map(&:hex).intersect?(@south_hexes)
         end
 
         def check_other(route)
-          return if g_train?(route.train)
-          return unless route.stops.map(&:hex).intersect?(port_hexes)
+          raise GameError, 'All Goods Train routes must include a port and a mine' if g_train?(route.train) &&
+                                                                                      (!port_hexes?(route) ||
+                                                                                      !mine_hexes?(route))
 
-          raise GameError, 'Only G trains can run to ports'
+          raise GameError, 'Only Goods Trains can run to ports' if !g_train?(route.train) && port_hexes?(route)
         end
 
         def revenue_for(route, stops)
           revenue = super
           train = route.train
           corp = route.corporation
 
-          revenue += 10 if corp.assigned?(p4_company) && stops.any? { |s| s.hex.id == C18_HEX }
+          revenue += north_south_bonus_check(route, train)[:revenue]
+          revenue += bonus_for_p4_company(corp, stops)
 
-          { p3_company => F3_PORT, p8_company => F3_PORT, p9_company => H9_PORT, p12_company => G6_PORT }.each do |company, port|
-            revenue += (company == p12_company ? 20 : 10) if g_train?(train) && corp.assigned?(company) && stops.any? do |s|
-                                                               s.hex.id == port
-                                                             end
+          if g_train?(train)
+            revenue += port_mine_bonus_check(route, train)[:revenue]
+            revenue += bonus_for_port_companies(corp, stops)
           end
 
-          revenue += north_south_bonus(stops)[:revenue]
-          revenue += port_mine_bonus(stops)[:revenue]
-
           revenue
         end
 
-        def north_south_bonus(stops)
-          bonus = { revenue: 0 }
+        def bonus_for_p4_company(corp, stops)
+          return 10 if corp.assigned?(p4_company) && stops.any? { |s| s.hex.id == C18_HEX }
```

Comment:
The `assigned?` isn't right, that is for AssignCorporation abilities, which are generally ones where the ability can be used when the private company is owned by a player, who chooses which of their corporation gets to use the ability.

Can this be changed to give P4 (and the other privates with revenue bonuses) a HexBonus ability? You'd then have something like this code to work out the bonuses:
https://github.com/tobymao/18xx/blob/1c6e10f63f6889c61c6fe61493d5f1572d042e83/lib/engine/game/g_1867/game.rb#L681-L685
I don't think you'd need to add any extra tests to the loop for goods trains for the non-P4 privates, as (if I understand this correctly) they give bonuses for ports, and it isn't possible for non-goods trains to run to ports.

Reply:
Or `revenue += ability.amount if ability.hexes.intersect?(stops.map(&:hex).map(&:id))`

Reply:
done